File: SongWei_20170209/piratecat_net/Controls/Chart/Service/UserSessionServiceEx.cs
/***********************************************************************************\
*                                                                                   *
* UserSessionServiceEx.cs - User session service functions, types, and definitions. *
*                                                                                   *
*               Version 1.00  ★★★                                                *
*                                                                                   *
*               Copyright (c) 2016-2016, Piratecat. All rights reserved.            *
*               Created by Lord 2016/1/20.                                          *
*                                                                                   *
************************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;
using OwLib;
using OwLibCT;

namespace piratecat
{
    /// <summary>
    /// 用户会话服务扩展
    /// </summary>
    public class UserSessionServiceEx : UserSessionService
    {
        #region Lord 2016/10/4
        /// <summary>
        /// 创建用户会话服务扩展
        /// </summary>
        public UserSessionServiceEx()
        {
        }

        /// <summary>
        /// 会话列表
        /// </summary>
        private Dictionary<String, UserSession> m_sessions = new Dictionary<String, UserSession>();

        private int m_getListRequestID = BaseService.GetRequestID();

        /// <summary>
        /// 获取列表的请求ID
        /// </summary>
        public int GetListRequestID
        {
            get { return m_getListRequestID; }
        }

        private bool m_loaded;

        /// <summary>
        /// 获取或设置是否已加载
        /// </summary>
        public bool Loaded
        {
            get { return m_loaded; }
        }

        private int m_operatorRequestID = BaseService.GetRequestID();

        /// <summary>
        /// 获取操作请求ID
        /// </summary>
        public int OperatorRequestID
        {
            get { return m_operatorRequestID; }
        }

        /// <summary>
        /// 销毁对象
        /// </summary>
        public override void Dispose()
        {
            if (!IsDisposed)
            {
                m_sessions.Clear();
                base.Dispose();
            }
        }

        /// <summary>
        /// 获取会话
        /// </summary>
        /// <param name="key">键</param>
        /// <param name="session">会话</param>
        /// <returns>状态</returns>
        public int GetSession(String key, ref UserSession session)
        {
	        int state = 0;
            lock(m_sessions)
            {
                if(m_sessions.ContainsKey(key))
                {
                    session = m_sessions[key];
                    state=1;
                }
            }
            return state;
        }

        public override void OnReceive(CMessage message)
        {
            List<UserSession> sessions = new List<UserSession>();
            GetSessions(sessions, message.m_body, message.m_bodyLength);
            int sessionsSize = sessions.Count;
            switch (message.m_functionID)
            {
                case FUNCTIONID_USERSESSION_GETSESSIONS:
                    {
                        lock (m_sessions)
                        {
                            for (int i = 0; i < sessionsSize; i++)
                            {
                                UserSession session = sessions[i];
                                m_sessions[session.m_key] = session;
                                OnSessionChanged(session);
                            }
                        }
                        m_loaded = true;
                        break;
                    }
                case FUNCTIONID_USERSESSION_ADDSESSION:
                    {
                        lock (m_sessions)
                        {
                            for (int i = 0; i < sessionsSize; i++)
                            {
                                UserSession session = sessions[i];
                                m_sessions[session.m_key] = session;
                                OnSessionChanged(session);
                            }
                        }
                        break;
                    }
                case FUNCTIONID_USERSESSION_DELETESESSION:
                    {
                        lock (m_sessions)
                        {
                            for (int i = 0; i < sessionsSize; i++)
                            {
                                UserSession session = sessions[i];
                                if (m_sessions.ContainsKey(session.m_key))
                                {
                                    m_sessions.Remove(session.m_key);

                                    session.m_value = "";
                                    OnSessionChanged(session);
                                }
                            }
                            break;
                        }
                    }
            }
            base.OnReceive(message);
        }

        /// <summary>
        /// 会话改变方法
        /// </summary>
        /// <param name="session">会话</param>
        public virtual void OnSessionChanged(UserSession session)
        {
        }
        #endregion
    }
}

File: SongWei_20170209/piratecat/piratecat/Chart/Service/UserSessionServiceEx.cpp
#include "..\\..\\stdafx.h"
#include "UserSessionServiceEx.h"

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

UserSessionServiceEx::UserSessionServiceEx()
{
	m_getListRequestID = BaseService::GetRequestID();
	m_loaded = false;
	m_operatorRequestID = BaseService::GetRequestID();
}

UserSessionServiceEx::~UserSessionServiceEx()
{
	m_lock.Lock();
	m_sessions.clear();
	m_lock.UnLock();
}

int UserSessionServiceEx::GetListRequestID()
{
	return m_getListRequestID;
}

bool UserSessionServiceEx::IsLoaded()
{
	return m_loaded;
}

int UserSessionServiceEx::GetOperatorRequestID()
{
	return m_operatorRequestID;
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

int UserSessionServiceEx::GetSession(String key, UserSession *session)
{
	int state = 0;
	m_lock.Lock();
	map<String, UserSession>::iterator sIter  = m_sessions.find(key);
	if(sIter != m_sessions.end())
	{
		*session = sIter->second;
		state = 1;
	}
	m_lock.UnLock();
	return state;
}

void UserSessionServiceEx::OnReceive(CMessage *message)
{
    vector<UserSession> sessions;
	UserSessionService::GetSessions(&sessions, message->m_body, message->m_bodyLength);
    int sessionsSize = (int)sessions.size();
	switch (message->m_functionID)
    {
		case FUNCTIONID_USERSESSION_GETSESSIONS:
			{
				m_lock.Lock();
				for (int i = 0; i < sessionsSize; i++)
				{
					UserSession session = sessions[i];
					m_sessions[session.m_key] = session;
					OnSessionChanged(&session);
				}
				m_lock.UnLock();
				m_loaded = true;
				break;
			}
        case FUNCTIONID_USERSESSION_ADDSESSIONS:
            {
				m_lock.Lock();
                for (int i = 0; i < sessionsSize; i++)
                {
					UserSession session = sessions[i];
					m_sessions[session.m_key] = session;
					OnSessionChanged(&session);
                }
				m_lock.UnLock();
                break;
            }
        case FUNCTIONID_USERSESSION_DELETESESSIONS:
            {
				m_lock.Lock();
                for (int i = 0; i < sessionsSize; i++)
                {
					UserSession session = sessions[i];
					map<String, UserSession>::iterator sIter  = m_sessions.find(session.m_key);
					if(sIter != m_sessions.end())
					{
						m_sessions.erase(sIter);
					}
					session.m_value = L"";
					OnSessionChanged(&session);
                }
				m_lock.UnLock();
                break;
            }
    }
	UserSessionService::OnReceive(message);
}

void UserSessionServiceEx::OnSessionChanged(UserSession *session)
{
}
